public/scripts/openai.js

```diff
@@ -299,199 +319,126 @@ function formatWorldInfo(value) {
 }
 
 async function prepareOpenAIMessages(name2, storyString, worldInfoBefore, worldInfoAfter, extensionPrompt, bias, type, quietPrompt) {
-    const isImpersonate = type == "impersonate";
-    let this_max_context = oai_settings.openai_max_context;
-    let enhance_definitions_prompt = "";
-    let nsfw_toggle_prompt = oai_settings.nsfw_toggle ? oai_settings.nsfw_prompt : oai_settings.nsfw_avoidance_prompt;
-
-    // Experimental but kinda works
-    if (oai_settings.enhance_definitions) {
-        enhance_definitions_prompt = "If you have more knowledge of " + name2 + ", add to the character's lore and personality to enhance them but keep the Character Sheet's definitions absolute.";
+    const chatCompletion = promptManager.getChatCompletion();
+
+    // Prepare messages
+    const worldInfoBeforeMessage = chatCompletion.makeSystemMessage(formatWorldInfo(worldInfoBefore));
+    const worldInfoAfterMessage = chatCompletion.makeSystemMessage(formatWorldInfo(worldInfoAfter));
+    const characterInfoMessages = chatCompletion.makeSystemMessage(substituteParams(storyString));
+    const newChatMessage = chatCompletion.makeSystemMessage('[Start new chat]');
+    const chatMessages = openai_msgs;
+    const biasMessage = chatCompletion.makeSystemMessage(bias.trim());
+
+    // Prepare context
+    chatCompletion
+        .replace('worldInfoBefore', worldInfoBeforeMessage)
+        .replace('worldInfoAfter', worldInfoAfterMessage)
+        .replace('characterInfo', characterInfoMessages)
+        .replace('newExampleChat', newChatMessage)
+        .replace('newMainChat', newChatMessage)
+        .replace('chatHistory', chatMessages)
+
+    // Handle group chats
+    if (selected_group) {
+        const names = getGroupMembers(groups);
+        const groupChatMessage = chatCompletion.makeSystemMessage(`[Start a new group chat. Group members: ${names}]`);
+        const groupNudgeMessage = chatCompletion.makeSystemMessage(`[Write the next reply only as ${name2}]`);
+        chatCompletion.replace('newMainChat', groupChatMessage)
+        chatCompletion.insertAfter('main', 'groupNudgeMessage', groupNudgeMessage);
+        chatCompletion.remove('main'); // ToDo: Instead of removing, the main prompt could be altered to reinforce the situation being a group scene.
```

Comment:
The main prompt creates a contradiction with groupChatMessage. With a dialogue between one user and one character being established despite it being a group scene. Though, you're right that the main prompt should be sent and the impact of the contradiction is probably negligible. 

Alternatively, we could send the main prompt but repalce {{char}} with the names of all participants, similiar to the nudge. The result would be: Write Aqua's next reply in a fictional chat between Aqua, Darkness, Megumin and TestUser [...] [Write the next reply only as Aqua]

Establishing an explicit scene and clear instruction.

Reply:
I tackled this and ran into the issue that multiple {{char}} variables can occur in a single prompt. Which is kinda obvious but i didn't think of it. Also it would be nice to let the user know what's happening instead of silently replacing {{char}} with the group names. 

To resolve this, i introduced two new variables: {{charIfNotGroup}} and {{group}}. Altering the main prompt to: 

Write {{char}}'s next reply in a fictional chat between {{charIfNotGroup}} and {{user}}. [...] 

This way it's at least understandable what happens and as a nice side effect it may enable users to compose custom prompts better suitable for groups, by using the two new variables.

Reply:
:+1: